Add non-eager registration to dispatch autogen

antoniojkim opened the pull request:
Proposed fix for #73919

Now, the TorchScript LTC backend lazily registers the lazy kernels on initialization instead of on dynamic library load.

- introduces a new 'eager_registration' parameter to the lazy codegen
- eager_registration=true (default) preserves the existing behavior: backend nativefunc impls are registered at static initializaiton time
- eager_registration=false is used for the in-tree lazy TS backend, such that it does not register any impls against the Lazy key until the ts bakcend init function is explicitly called 

CC: @wconstab @henrytwo @JackCaoG  @ezyang @bdhirsh

antoniojkim committed:
Add non-eager registration to dispatch autogen

antoniojkim committed:
Fix env in gen.py

antoniojkim committed:
Fix cpp_namespace

antoniojkim committed:
Properly initialize TS backend in LTC tests

antoniojkim committed:
Fix init namespace

antoniojkim commented:
@wconstab can I get a review for this PR?

wconstab commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -201,12 +204,13 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
 
     # Generate nativefunction declarations
     gen_dispatchkey_nativefunc_headers(fm, class_name, cpp_namespace, backend_indices,
```
could you add a comment explaining why eager_registration=False for lazy TS backend?

wconstab commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -208,7 +208,9 @@ def gen_dispatchkey_nativefunc_headers(
-        autograd_dispatch_key: Optional[DispatchKey]) -> None:
+        autograd_dispatch_key: Optional[DispatchKey],
+        backend_name: str = "",
+        eager_registration: bool = True) -> None:
```
could you add a comment about what 'eager_registration' means here?

wconstab commented on aten/src/ATen/templates/RegisterDispatchKey.cpp:
```diff
@@ -58,7 +67,15 @@ namespace {
+
+  };
+
+  #if EAGER_REGISTRATION
```
ok- so i think what happens is

- for external backends, EAGER_REGISTRATION=1 causes the module registration function to be called at the same time it used to be called, in other words, at c++ static initialization time.
- for torchscript backend, EAGER_..=0 makes the c++ static initializer simply install a function pointer, and then only upon calling init_ts_backend is it finally invoked.

Could you write a comment to this effect in the template if so?

wconstab commented on aten/src/ATen/templates/DispatchKeyNativeFunctions.h:
```diff
@@ -10,4 +10,11 @@ struct ${class_name} {
+#define EAGER_REGISTRATION ${eager_registration}
+
+#if !EAGER_REGISTRATION
+extern TORCH_API std::function<void(void)> Register${BackendName}${DispatchKey}Modules;
```
wait, why is this marked extern?  I thought that the function is defined in RegisterDispatchKey.cpp either way.

I think you can call it from init_ts_backend with just a TORCH_API macro on it?

antoniojkim replied:
Its marked `extern` as multiple source files include `LazyNativeFunctions` causing a multiple definition of the same symbol error. Marking it `extern` just makes the symbol known, but does not define it.

wconstab commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_eager_fallback.cpp:
```diff
@@ -55,8 +55,12 @@ void ltc_eager_fallback(const c10::OperatorHandle& op,
 TORCH_LIBRARY_IMPL(_, Lazy, m) {
-  m.fallback(torch::CppFunction::makeFromBoxedFunction<&ltc_eager_fallback>());
+  register_ts_ltc_eager_fallback = [&]() {
+    m.fallback(torch::CppFunction::makeFromBoxedFunction<&ltc_eager_fallback>());
```
this one doesn't need to be guarded under EAGER_REGISTRATION=1?

It's currently in a file that is only part of the TS backend, so that works for now.   I think we may want to share some of the fallback logic with xla / other backends later, but i guess we can just update the registration at that time.

antoniojkim replied:
Its guarded under `EAGER_REGISTRATION=1` as a different LTC backend may want to implement different fallback logic. Which wouldn't be possible if the TS backend registers the fallback at c++ static initialization time.

wconstab commented:
I think this looks good, I'll also show it to @JackCaoG as a heads up, but i'm pretty sure it won't impact them.

ezyang commented:
Is there anything specific you want from me on this PR?

wconstab commented:
@ezyang Just tagged you for visibilty, but it looked ok to me.

antoniojkim committed:
Address PR comments

antoniojkim commented:
@wconstab I've addressed your comments. Can I get another review on this? Ideally it'd be great if this change also landed in `master` in https://github.com/pytorch/pytorch/pull/74111

wconstab commented:
Yea, I am happy with it now.  I am not sure what the easiest landing path is, I am thinking of possibly landing it directly to master after my 'land ts backend' PR, but if it's blocking your work I can just land it on staging now.

antoniojkim commented:
> Yea, I am happy with it now. I am not sure what the easiest landing path is, I am thinking of possibly landing it directly to master after my 'land ts backend' PR, but if it's blocking your work I can just land it on staging now.

Its not exactly "blocking" my work as I've cherry picked the changes locally and am using it that way. However, getting it to land in `lazy_tensor_staging` sooner rather than later will probably be preferable. You can decide then whether the change should land in `master` alongside your "land TS backend" PR or in a follow up

antoniojkim commented:
@wconstab can you please merge this? I don't have write access
